### PR TITLE
Prefetch img, scripts and stylesheets during parsing

### LINKED ISSUE #3847: Off-main-thread, speculative HTML parsing
By running the HTML parser in a separate task from script, we could allow tokenization and speculative tree building to continue while content scripts are fetched and executed.

References on HTML parser threading in Gecko and WebKit:
- https://hsivonen.fi/speculative-html5-parsing/
- https://developer.mozilla.org/en-US/docs/Mozilla/Gecko/HTML_parser_threading
- https://bugs.webkit.org/show_bug.cgi?id=106127

### PR DESCRIPTION
**Pull request by asajeffrey**
Eagerly tokenize html input, and scan for `script` and `img` tags which can be prefetched into the cache. This allows resources to be loaded concurrently, which would otherwise be loaded sequentially due to being blocked by script execution. On the cloth webgl demo, this takes the time-to-paint down from 732ms to 560ms.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #3847
- [x] These changes do not require tests because it's a perf improvement

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/24148)

### CODE REVIEW

**Comment by asajeffrey** [comment]
IRC chat: https://mozilla.logbot.info/servo/20190904#c16590059 and https://mozilla.logbot.info/servo/20190905#c16593012

**Comment by asajeffrey** [comment]
One thing I'm not sure about is that we're issuing a GET when prefetching, even for content which isn't cacheable. We should probably first issue a HEAD to see if the content is cacheable. Perhaps this should be a follow-up issue?

File: components/script/dom/servoparser/prefetch.rs
Comment (jdm) [inline components/script/dom/servoparser/prefetch.rs]: Could we use the annotation where necessary instead?

File: components/script/dom/servoparser/prefetch.rs
Comment (jdm) [inline components/script/dom/servoparser/prefetch.rs]: I would prefer to derive JSTraceable on Tokenizer instead to avoid accidents in the future if we update that struct.

File: components/script/dom/servoparser/prefetch.rs
Comment (jdm) [inline components/script/dom/servoparser/prefetch.rs]: We should also handle the <base> tag and update the stored base URL.

File: components/script/dom/servoparser/prefetch.rs
Comment (jdm) [inline components/script/dom/servoparser/prefetch.rs]: Is it worth prefetching external stylesheets as well?

File: components/script/dom/servoparser/prefetch.rs
Comment (jdm) [inline components/script/dom/servoparser/prefetch.rs]: We should extract the operation for building a request for scripts/images/etc. from htmlscriptelement.rs and htmlimageelement.rs into reusable routines to ensure the prefetched request matches the actual request - we will need to account for CORS settings, referrer policies, the origin header, integrity metadata, etc.

**Review by jdm — CHANGES_REQUESTED** [review]
I like the simplicity of this model. We just need to make sure that we don't unintentionally bypass security features of the platform by prefetching more liberally than the actual request the page intends to make.

**Comment by asajeffrey** [comment]
I don't think there's security issues here, since the data requested by the prefetch is only used to populate the cache, and isn't exposed to script. If we do set the headers according to htmlscriptelement and htmlimageelement, then we do need to make sure we're not sending cookies we shouldn't.

**Comment by asajeffrey** [comment]
IRC chat at https://mozilla.logbot.info/servo/20190906#c16596185

**Comment by asajeffrey** [comment]
One thing I realized doing this is that fetch uses the origin, which is mutable, but for speculative prefetching we don't get much choice but to use the original origin.

File: components/script/dom/servoparser/prefetch.rs
Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: Oh this is weird, I replied to this comment but github appears to have lost the reply...

Anyway, the problem is that HtmlTokenizer doesn't implement `JSTraceable`, and it's difficult to do so as its generic, and the tokenizer API provides no way to access the sink. If it did, the implementation would be pretty easy:
```rust
    impl<Sink: JSTraceable> for HtmlTokenizer<Sink> {
        fn trace(&self) { self.sink().trace() } 
   } 
```
but there's no `sink()` method, so we can't really do that.

Comment (jdm) [reply components/script/dom/servoparser/prefetch.rs]: We have a JSTraceable implementation for the HTML tokenizer in script/dom/servoparser/html.rs. I recommend following its example.

Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: Er, but stripping out the unnecessary code leaves just:
```rust
#[allow(unsafe_code)]
unsafe impl JSTraceable for HtmlTokenizer<PrefetchSink>> {
    unsafe fn trace(&self, trc: *mut JSTracer) {
    }
}
```
which doesn't have obvious advantages over `unsafe_no_jsmanaged_fields!`.

Comment (jdm) [reply components/script/dom/servoparser/prefetch.rs]: Can't we still derive JStraceable on Sink and call `self.sink.trace()` at least?

Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: The problem is we don't have access to the sink, since the tokenizer doesn't expose an API to access it. We couuuld use an Rc, but that seems a bit round-the-houses.

Comment (jdm) [reply components/script/dom/servoparser/prefetch.rs]: Isn't the tokenizer [this type](https://docs.rs/html5ever/0.23.0/html5ever/tokenizer/struct.Tokenizer.html). It has a public `sink` field. What am I missing?

Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: Argh, I was looking for an accessor method, I didn't spot that the field was public. D'oh!

Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: Okay, fixed.

**Review by jdm — APPROVED** [review]

**Comment by asajeffrey** [comment]
```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"1\" but got \"2\"", 
    "stack": "count/xhr.onload<@http://web-platform.test:8000/html/semantics/document-metadata/the-link-element/document-without-browsing-context.html:13:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1611:25\nTest.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1651:32\n", 
    "subtest": "Create a stylesheet in innerHTML document", 
    "test": "/html/semantics/document-metadata/the-link-element/document-without-browsing-context.html", 
    "line": 244667, 
    "action": "test_result", 
    "expected": "PASS"
}
```
that test counts how may times a stylesheet is fetched, not too surprising it now fails!

**Comment by asajeffrey** [comment]
@jdm this is yours apparently... https://github.com/web-platform-tests/wpt/commit/42bfe56f51fc7bc909cd847706aa3de1b1775216

I'm trying to find the appropriate spec link for this behaviour. Do you remember?

**Comment by asajeffrey** [comment]
Ah, via https://github.com/servo/servo/pull/12241 I found https://github.com/whatwg/html/issues/1495.

**Comment by asajeffrey** [comment]
Hmm, encoding issues.

**Comment by asajeffrey** [comment]
IRC chat: https://mozilla.logbot.info/servo/20190909#c16601081

TL;dr @nox gets a gold star!

**Comment by asajeffrey** [comment]
Squashed.

@jdm r+d on IRC https://mozilla.logbot.info/servo/20190909#c16601719

@bors-servo r=jdm

**Comment by asajeffrey** [comment]
Hmm...
```
{
    "status": "CRASH", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/resource-timing/resource_script_types.html", 
    "line": 85130, 
    "action": "test_result", 
    "expected": "OK"
}
```

**Comment by asajeffrey** [comment]
The crash being
```

  ▶ CRASH [expected OK] /resource-timing/resource_script_types.html
  │ 
  │ _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
  │ 
  │ 
  │ 
  │ called `Result::unwrap()` on an `Err` value: InvalidUri(InvalidUriChar) (thread <unnamed>, at src/libcore/result.rs:1084)
  │ stack backtrace:
  │    0: backtrace::backtrace::trace::h78e71490c44ccd31 (0x102fa2cbc)
  │    1: backtrace::capture::Backtrace::new::hf7130b3c183a6059 (0x102fa1f48)
  │    2: servo::main::{{closure}}::h0997dd67f6a5f8e1 (0x1001515fc)
  │    3: std::panicking::rust_panic_with_hook::h099fe3773ce1a793 (0x10313a5b2)
  │    4: std::panicking::continue_panic_fmt::h8a4b76634870b383 (0x10313a01d)
  │    5: rust_begin_unwind (0x103139f19)
  │    6: core::panicking::panic_fmt::h885799c19d2b8b0c (0x10315a74f)
  │    7: core::result::unwrap_failed::h91161a3faccc6622 (0x10315a829)
  │    8: net::http_loader::http_network_or_cache_fetch::h353385e38e4e33e3 (0x10214618e)
  │    9: net::http_loader::http_fetch::h7bd22bab056e2000 (0x10213e117)
  │   10: net::fetch::methods::scheme_fetch::hc08b944cc9fee259 (0x102110ed5)
  │   11: net::fetch::methods::main_fetch::hdbb17379cf1a139b (0x10210f0c8)
  │   12: net::fetch::methods::fetch_with_cors_cache::h3de060f49d07cd70 (0x10210d425)
  │   13: net::fetch::methods::fetch::h237bf47dc58c2d3c (0x10210d284)
  │   14: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hf08e1de077cdd28f (0x10208a8fd)
  │   15: _ZN3std9panicking3try7do_call17hbb5aa698c9e153daE.llvm.14923142366686951073 (0x102168d9b)
  │   16: __rust_maybe_catch_panic (0x10314294f)
  │   17: _ZN77_$LT$rayon_core..job..HeapJob$LT$BODY$GT$$u20$as$u20$rayon_core..job..Job$GT$7execute17h8cef6d65682639eaE.llvm.16426940954949856000 (0x10203a604)
  │   18: rayon_core::registry::WorkerThread::wait_until_cold::h09146fa8a53f3089 (0x102b8696e)
  │   19: rayon_core::registry::ThreadBuilder::run::hefc1c1df25f49f05 (0x102b84f24)
  │   20: _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h4bd602cfd4eaa329E.llvm.12129492943617047840 (0x102b88b72)
  │   21: _ZN3std9panicking3try7do_call17hb8b5e489a31c1a54E.llvm.8371471668520928442 (0x102b89842)
  │   22: __rust_maybe_catch_panic (0x10314294f)
  │   23: core::ops::function::FnOnce::call_once{{vtable.shim}}::h7bcd519f61762799 (0x102b82bb2)
  │   24: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hb03304ace7f89c63 (0x103129cde)
  │   25: std::sys::unix::thread::Thread::new::thread_start::hd2d570f25bb299f3 (0x103141e7e)
  │   26: _pthread_body (0x7fff584bd661)
  │   27: _pthread_start (0x7fff584bd50d)
  │ [2019-09-09T20:56:45Z ERROR servo] called `Result::unwrap()` on an `Err` value: InvalidUri(InvalidUriChar)
  │ Rayon: detected unexpected panic; aborting
  │ Redirecting call to abort() to mozalloc_abort
  │ 
  │ Stack trace
  │ stack backtrace:
  │    0: backtrace::backtrace::trace::h78e71490c44ccd31 (0x102fa2cbc)
  │    1: backtrace::capture::Backtrace::new::hf7130b3c183a6059 (0x102fa1f48)
  │    2: servo::install_crash_handler::handler::ha349d978d5b25fe0 (0x10015092c)
  │    3: _sigtramp (0x7fff584b3f5a)
  │    4: mozalloc_abort (0x101c46331)
  │    5: abort (0x101c4634d)
  │    6: std::sys::unix::abort_internal::h536db4d09a71f8fd (0x103142359)
  │    7: std::process::abort::he114549223cf212e (0x1031363b9)
  │    8: rayon_core::registry::Registry::handle_panic::h45c24b1d1ddb663b (0x102b8647a)
  │    9: _ZN77_$LT$rayon_core..job..HeapJob$LT$BODY$GT$$u20$as$u20$rayon_core..job..Job$GT$7execute17h8cef6d65682639eaE.llvm.16426940954949856000 (0x10203a62c)
  │   10: rayon_core::registry::WorkerThread::wait_until_cold::h09146fa8a53f3089 (0x102b8696e)
  │   11: rayon_core::registry::ThreadBuilder::run::hefc1c1df25f49f05 (0x102b84f24)
  │   12: _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h4bd602cfd4eaa329E.llvm.12129492943617047840 (0x102b88b72)
  │   13: _ZN3std9panicking3try7do_call17hb8b5e489a31c1a54E.llvm.8371471668520928442 (0x102b89842)
  │   14: __rust_maybe_catch_panic (0x10314294f)
  │   15: core::ops::function::FnOnce::call_once{{vtable.shim}}::h7bcd519f61762799 (0x102b82bb2)
  │   16: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hb03304ace7f89c63 (0x103129cde)
  │   17: std::sys::unix::thread::Thread::new::thread_start::hd2d570f25bb299f3 (0x103141e7e)
  │   18: _pthread_body (0x7fff584bd661)
  └   19: _pthread_start (0x7fff584bd50d)
```

**Comment by asajeffrey** [comment]
OK, I think what's going on is that for some reason (on MacOS, in CI) we end up with a URL that Servo is happy about, but reparsing it as a hyperium URI fails. I removed some uses of `unwrap()` from `http_loader`, let's see if that fixes the issue.

@bors-servo try=wpt-mac

**Comment by jdm** [comment]
I would be really interested to find out what the invalid URI is, because it's not clear to me why prefetching should cause different behaviour in that test with respect to providing a referrer.

**Comment by asajeffrey** [comment]
It would indeed. I'm not too surprised prefetching can cause a problem for tests which use `document.write` and overwrite the original content of a document, since in that case there's resources being prefetched that are never used.

Annoyingly I can't replicate the issue at my end, the easiest way to find it would be to have a variant of this code that uses `expect` rather than `unwrap` so we can see which URI parse is failing, then run it through CI.

If we're really really lucky, this'll turn out to be an intermittent.

**Comment by asajeffrey** [comment]
OK, I pushed an experiment which panics with a message when it finds the URL that doesn't parse. If the current run succeeds, we can run it to find out which URL is causing the problem.

**Comment by asajeffrey** [comment]
@jdm: if this fixes things, are you OK with logging the error and soldiering on rather than `panic!`ing?

**Comment by jdm** [comment]
Not without understanding why it happens. I suspect panicking here is actually a good thing.

**Comment by asajeffrey** [comment]
OK, sweeping the errors under the rug fixed the problem: https://tools.taskcluster.net/groups/AcVYJ7akTDemtbQV5YWc_A/tasks/XBo13IdoQ32C2YIQYQe9CQ/details

Let's try the experimental version that panics with the URL that failed to parse... @bors-servo try=wpt-mac

**Comment by asajeffrey** [comment]
I opened up an issue about our strategy around panic, we can go and be nerd-snipey there. https://github.com/servo/servo/issues/24167

**Comment by asajeffrey** [comment]
Sigh, serves me right for being clever and making sure we don't accidentally check this in by deliberately failing test-tidy.

@bors-servo try=wpt-mac

**Comment by asajeffrey** [comment]
Double sigh. @bors-servo try=wpt-mac

**Comment by asajeffrey** [comment]
So our error URL is `http://"empty_script.js/?id=5\\%22` which is caused by the test not doing string escaping properly in a bit of the test that gets overwritten by `document.write()`.

I'm not surprised hyperium isn't too happy about that URI, I'm a bit surprised our url crate allowed it.

```
  ▶ CRASH [expected OK] /resource-timing/resource_script_types.html
  │ 
  │ 
  │ 
  │ 
  │ Failed to parse http://"empty_script.js/?id=5\\%22 as authority (thread <unnamed>, at components/net/http_loader.rs:968)
  │ stack backtrace:
  │    0: backtrace::backtrace::trace::h78e71490c44ccd31 (0x10e1cecac)
  │    1: backtrace::capture::Backtrace::new::hf7130b3c183a6059 (0x10e1cdf38)
  │    2: servo::main::{{closure}}::h0997dd67f6a5f8e1 (0x10b37d39c)
  │    3: std::panicking::rust_panic_with_hook::h099fe3773ce1a793 (0x10e366572)
  │    4: std::panicking::continue_panic_fmt::h8a4b76634870b383 (0x10e365fdd)
  │    5: std::panicking::begin_panic_fmt::hd8effa78bdafa918 (0x10e365f3e)
  │    6: net::http_loader::http_network_or_cache_fetch::h353385e38e4e33e3 (0x10d3721d0)
  │    7: net::http_loader::http_fetch::h7bd22bab056e2000 (0x10d36a167)
  │    8: net::fetch::methods::scheme_fetch::hc08b944cc9fee259 (0x10d33cfa5)
  │    9: net::fetch::methods::main_fetch::hdbb17379cf1a139b (0x10d33b198)
  │   10: net::fetch::methods::fetch_with_cors_cache::h3de060f49d07cd70 (0x10d3394f5)
  │   11: net::fetch::methods::fetch::h237bf47dc58c2d3c (0x10d339354)
  │   12: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hf08e1de077cdd28f (0x10d2b6b0d)
  │   13: _ZN3std9panicking3try7do_call17hbb5aa698c9e153daE.llvm.14923142366686951073 (0x10d394d8b)
  │   14: __rust_maybe_catch_panic (0x10e36e90f)
  │   15: _ZN77_$LT$rayon_core..job..HeapJob$LT$BODY$GT$$u20$as$u20$rayon_core..job..Job$GT$7execute17h8cef6d65682639eaE.llvm.12173953782475729147 (0x10d266514)
  │   16: rayon_core::registry::WorkerThread::wait_until_cold::h09146fa8a53f3089 (0x10ddb295e)
  │   17: rayon_core::registry::ThreadBuilder::run::hefc1c1df25f49f05 (0x10ddb0f14)
  │   18: _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h4bd602cfd4eaa329E.llvm.12129492943617047840 (0x10ddb4b62)
  │   19: _ZN3std9panicking3try7do_call17hb8b5e489a31c1a54E.llvm.8371471668520928442 (0x10ddb5832)
  │   20: __rust_maybe_catch_panic (0x10e36e90f)
  │   21: core::ops::function::FnOnce::call_once{{vtable.shim}}::h7bcd519f61762799 (0x10ddaeba2)
  │   22: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hb03304ace7f89c63 (0x10e355c9e)
  │   23: std::sys::unix::thread::Thread::new::thread_start::hd2d570f25bb299f3 (0x10e36de3e)
  │   24: _pthread_body (0x7fff76978661)
  │   25: _pthread_start (0x7fff7697850d)
  │ [2019-09-10T13:46:06Z ERROR servo] Failed to parse http://"empty_script.js/?id=5\\%22 as authority
  │ Rayon: detected unexpected panic; aborting
  │ Redirecting call to abort() to mozalloc_abort
  │ 
  │ Stack trace
  │ stack backtrace:
  │    0: backtrace::backtrace::trace::h78e71490c44ccd31 (0x10e1cecac)
  │    1: backtrace::capture::Backtrace::new::hf7130b3c183a6059 (0x10e1cdf38)
  │    2: servo::install_crash_handler::handler::ha349d978d5b25fe0 (0x10b37c6cc)
  │    3: _sigtramp (0x7fff7696ef5a)
  │    4: mozalloc_abort (0x10ce720d1)
  │    5: abort (0x10ce720ed)
  │    6: std::sys::unix::abort_internal::h536db4d09a71f8fd (0x10e36e319)
  │    7: std::process::abort::he114549223cf212e (0x10e362379)
  │    8: rayon_core::registry::Registry::handle_panic::h45c24b1d1ddb663b (0x10ddb246a)
  │    9: _ZN77_$LT$rayon_core..job..HeapJob$LT$BODY$GT$$u20$as$u20$rayon_core..job..Job$GT$7execute17h8cef6d65682639eaE.llvm.12173953782475729147 (0x10d26653c)
  │   10: rayon_core::registry::WorkerThread::wait_until_cold::h09146fa8a53f3089 (0x10ddb295e)
  │   11: rayon_core::registry::ThreadBuilder::run::hefc1c1df25f49f05 (0x10ddb0f14)
  │   12: _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h4bd602cfd4eaa329E.llvm.12129492943617047840 (0x10ddb4b62)
  │   13: _ZN3std9panicking3try7do_call17hb8b5e489a31c1a54E.llvm.8371471668520928442 (0x10ddb5832)
  │   14: __rust_maybe_catch_panic (0x10e36e90f)
  │   15: core::ops::function::FnOnce::call_once{{vtable.shim}}::h7bcd519f61762799 (0x10ddaeba2)
  │   16: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hb03304ace7f89c63 (0x10e355c9e)
  │   17: std::sys::unix::thread::Thread::new::thread_start::hd2d570f25bb299f3 (0x10e36de3e)
  │   18: _pthread_body (0x7fff76978661)
  └   19: _pthread_start (0x7fff7697850d)
```

**Comment by asajeffrey** [comment]
@jdm, given that there are always going to be mismatches between hyperium's URI parser and the url crate, I will propose again logging the parse error and moving ahead. I'll work out whether this is an issue with the url crate being too generous, or hyperium being too strict, and file an issue.

**Comment by asajeffrey** [comment]
OK, that was a CI failure, the WPT test suite passed!

**Comment by asajeffrey** [comment]
Now waiting for a review from @nox.

**Comment by asajeffrey** [comment]
IRC chat https://mozilla.logbot.info/servo/20190910#c16603991

**Comment by asajeffrey** [comment]
Squashed.

**Review by nox — CHANGES_REQUESTED** [review]
I'm a bit anxious about doing that tbh, but it looks ok apart from some nits.

File: components/script/dom/servoparser/prefetch.rs
Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: You are very keen on unit structs. Done.

File: components/script/dom/servoparser/prefetch.rs
Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: Done.

File: components/script/dom/servoparser/prefetch.rs
Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: Done.

File: components/script/dom/servoparser/prefetch.rs
Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: Yeah, but if we waited to find out if script did any of that, we wouldn't be able to prefetch at all.

File: components/script/dom/servoparser/prefetch.rs
Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: According to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base "There can be only one <base> element in a document." But yes, I can imagine this causing resources to be prefetched erroneously.

**Review by asajeffrey — COMMENTED** [review]
OK, I think I addressed everything.

File: components/script/dom/servoparser/prefetch.rs
Comment (asajeffrey) [reply components/script/dom/servoparser/prefetch.rs]: OK, the spec has a note about this (https://html.spec.whatwg.org/multipage/semantics.html#the-base-element). Fixed.

**Comment by asajeffrey** [comment]
@nox, you okay with the changes?

**Pull request merged**
